REGRESSION: Black banners on webpage, sometimes content doesn't fully render when tab switching.

mattwoodrow opened the pull request:
#### 9941ce56b7a17396d18b9328222046d86d300551
<pre>
REGRESSION: Black banners on webpage, sometimes content doesn&apos;t fully render when tab switching.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266429">https://bugs.webkit.org/show_bug.cgi?id=266429</a>
&lt;<a href="https://rdar.apple.com/119668502">rdar://119668502</a>&gt;

Reviewed by Tim Horton.

If we&apos;re forcing a full display due to the existing front buffer being volatile and purged, then
we also need to reset the dirty region to the full bounds of the layer.

This stops us attempting a copy-forward from the old (purged) front buffer, and also makes sure
we clip to the full bounds, not what the web process thought the dirty region would be.

When the front buffer is potentially volatile, the web process always synchronously waits
for the response of this code, so the dirty region on that side should be updated correctly
already.

* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:

Canonical link: <a href="https://commits.webkit.org/272069@main">https://commits.webkit.org/272069@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4092fab86aba96d978d0ea458164e59200ac730

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32180 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33032 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/27608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6469 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27535 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27346 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6619 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34369 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27792 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/27695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6780 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4898 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/8517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7225 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/7512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
```diff
@@ -123,8 +123,10 @@ void RemoteImageBufferSet::ensureBufferForDisplay(const ImageBufferSetPrepareBuf
-        if (previousState == SetNonVolatileResult::Empty)
+        if (previousState == SetNonVolatileResult::Empty) {
             needsFullDisplay = true;
+            inputData.dirtyRegion = IntRect { { }, expandedIntSize(m_logicalSize) };
```
slightly odd that needsFullDisplay doesn't imply this

mattwoodrow replied:
The joys of the current code is that we have to iterate all the layers doing the swapping first, so that we can get the buffer handles sent back the web process as soon as possible (since it'll block waiting for that message at the end of painting).

Then once we've done that, we iterate a second time, initiating all the copy-forward operations.

Since there's two separate functions run on the `RemoteImageBufferSet`, `needsFullDisplay` isn't visible to the latter. Expanding the dirty region is the easiest way to share that information (that I could think of).

hortont424 commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
```diff
@@ -123,8 +123,10 @@ void RemoteImageBufferSet::ensureBufferForDisplay(const ImageBufferSetPrepareBuf
+        }
     }
 
     if (!m_frontBuffer || !inputData.supportsPartialRepaint || isSmallLayerBacking(m_frontBuffer->parameters()))
```
also seems like these also imply a need for a layer-sized dirty region? maybe it happens elsewhere?

mattwoodrow replied:
It does! These are all things that the web process knew about in advance and has accounted for. Unfortunately we need to compute and check these things on both sides :(

Adding an assert that the dirtyRegion is already expanded appropriately would be good at least.

mattwoodrow committed:
REGRESSION: Black banners on webpage, sometimes content doesn't fully render when tab switching.
https://bugs.webkit.org/show_bug.cgi?id=266429
<rdar://119668502>

Reviewed by Tim Horton.

If we're forcing a full display due to the existing front buffer being volatile and purged, then
we also need to reset the dirty region to the full bounds of the layer.

This stops us attempting a copy-forward from the old (purged) front buffer, and also makes sure
we clip to the full bounds, not what the web process thought the dirty region would be.

When the front buffer is potentially volatile, the web process always synchronously waits
for the response of this code, so the dirty region on that side should be updated correctly
already.

* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.h:

Canonical link: https://commits.webkit.org/272069@main
